pystac_client/stac_io.py

```diff
@@ -1,177 +1,98 @@
-                                   json=parameters)
-    else:
-        request = requests.Request(method=method,
-                                   url=original_request.url,
-                                   headers=headers,
-                                   params=parameters)
-
-    return request
-
-
-def get_pages(
-    session: requests.Session,
-    request: requests.Request,
-    next_resolver: Optional[Callable] = simple_stac_resolver,
-) -> Iterator[dict]:
-    """
-
-    Parameters
-    ----------
-    session : requests.Session
-        requests library Session object
-    request : requests.Request
-        The initial request to start paging. Subsequent requests will be determined by the ``next_resolver``.
-    next_resolver : Callable
-        An callable that will be used to construct the request for the next page of results based on the ``"next"``
-        link from the previous page.
-    """
-    while True:
-        # Yield all items
-        page = make_request(session, request)
-        yield page
-
-        # Get the next link and make the next request
-        next_link = next((link for link in page.get('links', []) if link['rel'] == 'next'), None)
-        if next_link is None:
-            break
-        request = next_resolver(next_link, request)
+class StacApiIO(DefaultStacIO):
+    def __init__(self, headers: Optional[Dict] = None):
+        """Initialize class for API IO
+
+        Args:
+            headers : Optional dictionary of headers to include in all requests
+
+        Returns:
+            StacApiIO : StacApiIO instance
+        """
+        # TODO - this should super() to parent class
+        self.session = Session()
+        self.session.headers.update(headers or {})
+
+    def read_text(self,
+                  source: Union[str, Link],
+                  *args: Any,
+                  parameters: Optional[dict] = {},
+                  **kwargs: Any) -> str:
+        """Overwrites the default method for reading text from a URL or file to allow :class:`urllib.request.Request`
+        instances as input. This method also raises any :exc:`urllib.error.HTTPError` exceptions rather than catching
+        them to allow us to handle different response status codes as needed."""
+        if isinstance(source, str):
+            href = source
+            if bool(urlparse(href).scheme):
+                return self.request(href, *args, parameters=parameters, **kwargs)
+            else:
+                with open(href) as f:
+                    href_contents = f.read()
+                return href_contents
+        elif isinstance(source, Link):
+            link = source.to_dict()
+            href = link['href']
+            # get headers and body from Link and add to request from simple stac resolver
+            merge = bool(link.get('merge', False))
+
+            # If the link object includes a "method" property, use that. If not fall back to 'GET'.
+            method = link.get('method', 'GET')
```

Comment:
The link object is returned from the server so should return whatever was initially used. I'm not sure if it's required by stac-api-spec, but I think some servers may not be returning it for GET requests.  Will have to check on that again.

Reply:
Okay, yeah, that might be worth clarifying in the API spec itself. It would definitely make these implementations easier if we could rely on that being in the link.

Reply:
Okay, yeah, that might be worth clarifying in the API spec itself. It would definitely make these implementations easier if we could rely on that being in the link.